libobs/audio-monitoring/nix/pulseaudio-output.c

```diff
@@ -0,0 +1,487 @@
+	data->channels = channels;
+skip:
+	pulseaudio_signal(0);
+}
+
+static void pulseaudio_stop_playback(struct audio_monitor *monitor)
+{
+	if (monitor->stream) {
+		pa_stream_disconnect(monitor->stream);
+		pa_stream_unref(monitor->stream);
+		monitor->stream = NULL;
+	}
+
+	blog(LOG_INFO, "Stopped Monitoring in '%s'", monitor->device);
+	blog(LOG_INFO, "Got %"
+		PRIuFAST32
+		" packets with %"
+		PRIuFAST64
+		" frames",
+	     monitor->packets, monitor->frames);
+
+	monitor->packets = 0;
+	monitor->frames = 0;
+}
+
+static bool audio_monitor_init(struct audio_monitor *monitor,
+                               obs_source_t *source)
+{
+	pthread_mutex_init_value(&monitor->playback_mutex);
+
+	monitor->source = source;
+
+	const char *id = obs->audio.monitoring_device_id;
+	if (!id) {
+		return false;
+	}
+
+	if (source->info.output_flags & OBS_SOURCE_DO_NOT_SELF_MONITOR) {
+		obs_data_t *s = obs_source_get_settings(source);
+		const char *s_dev_id = obs_data_get_string(s, "device_id");
+		bool match = devices_match(s_dev_id, id);
+		obs_data_release(s);
+
+		if (match) {
+			monitor->ignore = true;
+			blog(LOG_INFO, "Prevented feedback-loop in '%s'",
+			     s_dev_id);
+			return true;
+		}
+	}
+
+	pulseaudio_init();
+
+	if (strcmp(id, "default") == 0) {
+		get_default_id(&monitor->device);
+	} else {
+		monitor->device = bstrdup(id);
+	}
+
+	if (!monitor->device)
+		return false;
+
+	if (pulseaudio_get_server_info(pulseaudio_server_info,
+	                               (void *) monitor) < 0) {
+		blog(LOG_ERROR, "Unable to get server info !");
+		return false;
+	}
+
+	if (pulseaudio_get_source_info(pulseaudio_source_info, monitor->device,
+	                               (void *) monitor) < 0) {
+		blog(LOG_ERROR, "Unable to get source info !");
+		return false;
+	}
+	if (monitor->format == PA_SAMPLE_INVALID) {
+		blog(LOG_ERROR,
+		     "An error occurred while getting the source info!");
+		return false;
+	}
+
+	pa_sample_spec spec;
+	spec.format = monitor->format;
+	spec.rate = (uint32_t) monitor->samples_per_sec;
+	spec.channels = monitor->channels;
+
+	if (!pa_sample_spec_valid(&spec)) {
+		blog(LOG_ERROR, "Sample spec is not valid");
+		return false;
+	}
+
+	const struct audio_output_info *info = audio_output_get_info(
+		obs->audio.audio);
+
+	struct resample_info from = {
+		.samples_per_sec = info->samples_per_sec,
+		.speakers = info->speakers,
+		.format = AUDIO_FORMAT_FLOAT_PLANAR
+	};
+	struct resample_info to = {
+		.samples_per_sec = (uint32_t) monitor->samples_per_sec,
+		.speakers = pulseaudio_channels_to_obs_speakers(
+			monitor->channels),
+		.format = pulseaudio_to_obs_audio_format(monitor->format)
+	};
+
+	monitor->resampler = audio_resampler_create(&to, &from);
+	if (!monitor->resampler) {
+		blog(LOG_WARNING, "%s: %s", __FUNCTION__,
+		     "Failed to create resampler");
+		return false;
+	}
+
+	monitor->bytes_per_channel = get_audio_bytes_per_channel
+		(pulseaudio_to_obs_audio_format(monitor->format));
```

Comment:
You should probably keep the opening parenthesis here attached to the function call above.